code/modules/mob/emote.dm

```diff
@@ -1,56 +1,52 @@
-// All mobs should have custom emote, really..
-/mob/proc/custom_emote(var/m_type=1,var/message = null)
-
-
-	if(stat || !use_me && usr == src)
-		to_chat(usr, "You are unable to emote.")
-		return
-
-	var/muzzled = istype(src.wear_mask, /obj/item/clothing/mask/muzzle)
-	if(m_type == 2 && muzzled)
+//The code execution of the emote datum is located at code/datums/emotes.dm
+/mob/proc/emote(act, m_type = null, message = null)
+	act = lowertext(act)
+	var/param = message
+	var/custom_param = findtext(act, " ") // Someone was given as a parameter
+	if(custom_param)
+		param = copytext(act, custom_param + 1, length(act) + 1)
+		act = copytext(act, 1, custom_param)
+
+	var/datum/emote/E
+	E = E.emote_list[act]
+	if(!E)
+		to_chat(src, "<span class='notice'>Unusable emote '[act]'. Say *help for a list.</span>")
 		return
-
-	var/input
-	if(!message)
-		input = copytext(sanitize(input(src,"Choose an emote to display.") as text|null),1,MAX_MESSAGE_LEN)
-	else
-		input = message
-	if(input)
-		message = "<B>[src]</B> [input]"
-	else
-		return
-
-
-	if (message)
-		log_emote("[name]/[key] (@[x],[y],[z]): [message]")
-
- //Hearing gasp and such every five seconds is not good emotes were not global for a reason.
- // Maybe some people are okay with that.
-
-		for(var/mob/M in player_list)
-			if (!M.client)
-				continue //skip monkeys and leavers
-			if (istype(M, /mob/new_player))
-				continue
-			if(findtext(message," snores.")) //Because we have so many sleeping people.
-				break
-			if(M.stat == DEAD && M.client && M.client.prefs && (M.client.prefs.toggles & CHAT_GHOSTSIGHT) && !(M in viewers(src,null)))
-				M.show_message(message)
-
-
-		// Type 1 (Visual) emotes are sent to anyone in view of the item
-		if (m_type & 1)
-			visible_message(message)
-
-		// Type 2 (Audible) emotes are sent to anyone in hear range
-		// of the *LOCATION* -- this is important for pAIs to be heard
-		else if (m_type & 2)
-			for(var/mob/M in get_hearers_in_view(7, src))
-				M.show_message(message, m_type)
+	E.run_emote(src, param, m_type)
+
+/datum/emote/flip
+	key = "flip"
+	key_third_person = "flips"
+	restraint_check = TRUE
+	mob_type_allowed_typecache = list(/mob/living, /mob/dead/observer)
```

Comment:
we don't have typecaches though

Reply:
TG nerds have it
I'll do a mass search and replace if it's needed